### PR TITLE
Curvefs-tools: show topology

### LINKED ISSUE #879: curvefs-tool: list-topology
**Is your feature request related to a problem? (你需要的功能是否与某个问题有关?)**

show the topology of cluster

**Describe the solution you'd like (描述你期望的解决方法)**

**Describe alternatives you've considered (描述你想到的折衷方案)**

**Additional context/screenshots (更多上下文/截图)**
image

### PR DESCRIPTION
**Pull request by Cyber-SiKu**
### What problem does this PR solve?

Issue Number: close #879  

Problem Summary:

show the topology of cluster 

### What is changed and how it works?

What's Changed:

1. add list-toplogy rpc in mds/topology
2. add list-topology in curvefs_tool
3. add some function about topology in mds/topology

How it Works:

2. add list-topology in curvefs_tool, list-topology send list-topologyrequest to mds
 
Side effects(Breaking backward compatibility? Performance regression?): 

### Check List

- [ ] Relevant documentation/comments is changed or added
- [ ] I acknowledge that all my contributions will be made under the project's license

### CODE REVIEW

**Comment by Cyber-SiKu** [comment]
recheck

**Comment by YunhuiChen** [comment]
recheck

**Comment by Cyber-SiKu** [comment]
recheck

**Comment by Cyber-SiKu** [comment]
recheck

File: curvefs/src/tools/README.md [hunk 1]
```diff
@@ -65,6 +65,38 @@ Output:
...
+
+```Shell
+curvefs_tool status
+```
+
+Output:
+```shell
+mds version: 1.0.0
+leader mds: 10.182.2.45:36700
+standby mds: [ 10.182.2.45:16700 10.182.2.45:26700 ].
```
Comment (SeanHai) [inline curvefs/src/tools/README.md:82]: Mask the real ip.

File: curvefs/src/mds/topology/topology_manager.cpp [hunk 2]
```diff
@@ -1064,6 +1064,102 @@ void TopologyManager::GetMetaServersSpace(
...
+    ClusterInformation info;
+    if (topology_->GetClusterInfo(&info)) {
+        response->set_clusterid(info.clusterId);
+    } else {
+        response->set_clusterid("unknown");
+    }
+
+    // pool
+    auto poolInfos = new ListPoolResponse();
+    ListPool(nullptr, poolInfos);
```
Comment (SeanHai) [inline curvefs/src/mds/topology/topology_manager.cpp:1078]: ListPool(nullptr, response->mutable_pools())?

File: curvefs/src/mds/topology/topology_manager.cpp [hunk 3]
```diff
@@ -1064,6 +1064,102 @@ void TopologyManager::GetMetaServersSpace(
...
+    response->set_allocated_pools(poolInfos);
+    // zone
+    auto zone = new ListZoneResponse();
+    ListZone(zone);
```
Comment (SeanHai) [inline curvefs/src/mds/topology/topology_manager.cpp:1082]: Ditto

File: curvefs/src/tools/list/curvefs_topology_list.h [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+     * will fill serverId2ServerInfo's metaserverId list
+     *
+     * @return true
+     * @return false
+     * @details
+     */
+    bool GetMetaserverInfoFromResponse();
+
+ protected:
+    std::string clusterId_;
```
Comment (SeanHai) [inline curvefs/src/tools/list/curvefs_topology_list.h:121]: Maybe separate the variables and functions is more clearer.

File: curvefs/src/tools/list/curvefs_topology_list.cpp [hunk 5]
```diff
@@ -0,0 +1,318 @@
...
+                ret = false;
+            }
+        }
+    }
+    return ret;
+}
+
+void TopologyListTool::ShowPoolInfo(const PoolInfoType& pool) const {
+    auto const& poolInfo = pool.first;
+    std::cout << "poolId:" << poolInfo.poolid()
```
Comment (SeanHai) [inline curvefs/src/tools/list/curvefs_topology_list.cpp:260]: Maybe use a stringbuf and cout once is better.

File: curvefs/src/tools/list/curvefs_topology_list.cpp [hunk 6]
```diff
@@ -0,0 +1,318 @@
...
+    std::cout << "zoneList:{ ";
+    for (auto const& i : zoneIdVec) {
+        std::cout << i << " ";
+    }
+    std::cout << "}" << std::endl;
+}
+
+void TopologyListTool::ShowZoneInfo(const ZoneInfoType& zone) const {
+    auto const& zoneInfo = zone.first;
+    std::cout << "zoneId:" << zoneInfo.zoneid()
```
Comment (SeanHai) [inline curvefs/src/tools/list/curvefs_topology_list.cpp:274]: Ditto

File: curvefs/src/tools/list/curvefs_topology_list.cpp [hunk 7]
```diff
@@ -0,0 +1,318 @@
...
+    std::cout << "serverList:{ ";
+    for (auto const& i : serverIdVec) {
+        std::cout << i << " ";
+    }
+    std::cout << "}" << std::endl;
+}
+
+void TopologyListTool::ShowServerInfo(const ServerInfoType& server) const {
+    auto const& serverInfo = server.first;
+    std::cout << "serverId:" << serverInfo.serverid()
```
Comment (SeanHai) [inline curvefs/src/tools/list/curvefs_topology_list.cpp:287]: Ditto

File: curvefs/src/tools/README.md (hunk 1 above)
Comment (Cyber-SiKu) [reply curvefs/src/tools/README.md:82]: > Mask the real ip.

fix

File: curvefs/src/mds/topology/topology_manager.cpp (hunk 2 above)
Comment (Cyber-SiKu) [reply curvefs/src/mds/topology/topology_manager.cpp:1078]: > ListPool(nullptr, response->mutable_pools())

fix

File: curvefs/src/mds/topology/topology_manager.cpp (hunk 3 above)
Comment (Cyber-SiKu) [reply curvefs/src/mds/topology/topology_manager.cpp:1082]: > Ditto

fix

File: curvefs/src/tools/list/curvefs_topology_list.h (hunk 4 above)
Comment (Cyber-SiKu) [reply curvefs/src/tools/list/curvefs_topology_list.h:121]: > Maybe separate the variables and functions is more clearer.

fix

File: curvefs/src/tools/list/curvefs_topology_list.cpp (hunk 5 above)
Comment (Cyber-SiKu) [reply curvefs/src/tools/list/curvefs_topology_list.cpp:260]: > Maybe use a stringbuf and cout once is better.

fix

File: curvefs/src/tools/list/curvefs_topology_list.cpp (hunk 7 above)
Comment (Cyber-SiKu) [reply curvefs/src/tools/list/curvefs_topology_list.cpp:287]: > Ditto

fxi

**Comment by Cyber-SiKu** [comment]
recheck

**Comment by Cyber-SiKu** [comment]
recheck

**Comment by Cyber-SiKu** [comment]
recheck

File: curvefs/src/mds/topology/topology_manager.cpp [hunk 8]
```diff
@@ -1064,6 +1064,94 @@ void TopologyManager::GetMetaServersSpace(
...
+    }
+
+    // pool
+    ListPool(nullptr, response->mutable_pools());
+    // zone
+    ListZone(response->mutable_zones());
+    // Server
+    ListServer(response->mutable_servers());
+    // MetaServer
+    ListMetaserverOfCluster(response->mutable_metaservers());
```
Comment (wu-hanqing) [inline curvefs/src/mds/topology/topology_manager.cpp:1083]: comments in here are meaningless

File: curvefs/src/mds/topology/topology_manager.cpp [hunk 9]
```diff
@@ -1064,6 +1064,94 @@ void TopologyManager::GetMetaServersSpace(
...
+        if (topology_->GetZone(zoneId, &zone)) {
+            auto zoneInfo = response->add_zoneinfos();
+            zoneInfo->set_zoneid(zone.GetId());
+            zoneInfo->set_zonename(zone.GetName());
+            zoneInfo->set_poolid(zone.GetPoolId());
+        } else {
+            LOG(ERROR) << "Topology has counter an internal error: "
+                       << "[func:] ListZone, "
+                       << "[msg:] Zone not found, id = " << zoneId;
+            response->set_statuscode(TopoStatusCode::TOPO_INTERNAL_ERROR);
```
Comment (wu-hanqing) [inline curvefs/src/mds/topology/topology_manager.cpp:1100]: you didn't set response ok in the if branch

File: curvefs/src/mds/topology/topology_manager.cpp [hunk 10]
```diff
@@ -1064,6 +1064,94 @@ void TopologyManager::GetMetaServersSpace(
...
+            serverInfo->set_internalport(server.GetInternalPort());
+            serverInfo->set_externalip(server.GetExternalHostIp());
+            serverInfo->set_externalport(server.GetExternalPort());
+            serverInfo->set_zoneid(server.GetZoneId());
+            serverInfo->set_poolid(server.GetPoolId());
+        } else {
+            LOG(ERROR) << "Topology has counter an internal error: "
+                       << "[func:] ListServer, "
+                       << "[msg:] Server not found, id = " << serverId;
+            response->set_statuscode(TopoStatusCode::TOPO_INTERNAL_ERROR);
```
Comment (wu-hanqing) [inline curvefs/src/mds/topology/topology_manager.cpp:1124]: ditto

File: curvefs/src/mds/topology/topology_manager.cpp [hunk 11]
```diff
@@ -1064,6 +1064,94 @@ void TopologyManager::GetMetaServersSpace(
...
+            msInfo->set_port(ms.GetInternalPort());
+            msInfo->set_externalip(ms.GetExternalHostIp());
+            msInfo->set_externalport(ms.GetExternalPort());
+            msInfo->set_onlinestate(ms.GetOnlineState());
+            msInfo->set_serverid(ms.GetServerId());
+        } else {
+            LOG(ERROR) << "Topology has counter an internal error: "
+                       << "[func:] ListMetaServerOfCluster, "
+                       << "[msg:] metaserver not found, id = " << id;
+            response->set_statuscode(TopoStatusCode::TOPO_INTERNAL_ERROR);
```
Comment (wu-hanqing) [inline curvefs/src/mds/topology/topology_manager.cpp:1149]: ditto

File: curvefs/src/mds/topology/topology_service.cpp [hunk 12]
```diff
@@ -732,6 +732,24 @@ void TopologyServiceImpl::StatMetadataUsage(
     return;
 }
 
+void TopologyServiceImpl::ListTopology(
+    ::google::protobuf::RpcController* controller,
+    const ListTopologyRequest* request, ListTopologyResponse* response,
+    ::google::protobuf::Closure* done) {
+        brpc::ClosureGuard done_guard(done);
```
Comment (wu-hanqing) [inline curvefs/src/mds/topology/topology_service.cpp:739]: align

File: curvefs/src/tools/README.md [hunk 13]
```diff
@@ -65,6 +65,38 @@ Output:
...
+
+```Shell
+curvefs_tool status
+```
+
+Output:
+```shell
+mds version: 1.0.0
+leader mds: ***.***.***.***:36700
+standby mds: [ ***.***.***.***:16700 ***.***.***.***:26700 ].
```
Comment (wu-hanqing) [inline curvefs/src/tools/README.md:82]: use fake IP? like 192.168.1.1

File: curvefs/src/tools/README.md [hunk 14]
```diff
@@ -210,27 +242,58 @@ copysetInfo {
...
-    address: "10.182.2.45:16701:0"
+    address: "***.***.***.***:1****:0"
   }
 }
 
 ...
 
 ```
 
+### **list-topology**
```
Comment (wu-hanqing) [inline curvefs/src/tools/README.md:266]: TOC miss reference to this section

File: curvefs/src/tools/curvefs_tool_define.cpp [hunk 15]
```diff
@@ -201,5 +203,48 @@ std::string CopysetOpStatus2Str(
     return ret.str();
 }
 
+std::string PoolInfo2Str(const mds::topology::PoolInfo& poolInfo) {
+    std::stringstream ret;
+    ret << "poolId:" << poolInfo.poolid()
+        << ", poolName:" << poolInfo.poolname()
+        << ", createTime:" << poolInfo.createtime() << ", policy:{ "
+        << list::PoolPolicy(poolInfo.redundanceandplacementpolicy()) << " }, ";
```
Comment (wu-hanqing) [inline curvefs/src/tools/curvefs_tool_define.cpp:211]: why an extra ", " in the end?

File: curvefs/src/tools/status/curvefs_status_base_tool.h [hunk 16]
```diff
@@ -69,7 +69,7 @@ class StatusBaseTool : public CurvefsToolMetric {
     std::string version_;
     std::string hostType_;
     std::string versionSubUri_ = "";
-    std::string StatusSubUri_ = "";
+    std::string statusSubUri_ = "";
```
Comment (wu-hanqing) [inline curvefs/src/tools/status/curvefs_status_base_tool.h:72]: why some members have a default value and other not?

File: curvefs/src/mds/topology/topology_manager.cpp (hunk 9 above)
Comment (Cyber-SiKu) [reply curvefs/src/mds/topology/topology_manager.cpp:1100]: has set in start

File: curvefs/src/mds/topology/topology_manager.cpp (hunk 10 above)
Comment (Cyber-SiKu) [reply curvefs/src/mds/topology/topology_manager.cpp:1124]: > ditto

ditto

File: curvefs/src/mds/topology/topology_manager.cpp (hunk 11 above)
Comment (Cyber-SiKu) [reply curvefs/src/mds/topology/topology_manager.cpp:1149]: > ditto

ditto

File: curvefs/src/mds/topology/topology_manager.cpp (hunk 8 above)
Comment (Cyber-SiKu) [reply curvefs/src/mds/topology/topology_manager.cpp:1083]: > comments in here are meaningless

fix

File: curvefs/src/mds/topology/topology_service.cpp (hunk 12 above)
Comment (Cyber-SiKu) [reply curvefs/src/mds/topology/topology_service.cpp:739]: > align

fix

File: curvefs/src/tools/README.md (hunk 13 above)
Comment (Cyber-SiKu) [reply curvefs/src/tools/README.md:82]: > 192.168.1.1

fix

File: curvefs/src/tools/README.md (hunk 14 above)
Comment (Cyber-SiKu) [reply curvefs/src/tools/README.md:266]: > TOC miss reference to this section

TOC only for 2rd section

File: curvefs/src/tools/status/curvefs_status_base_tool.h (hunk 16 above)
Comment (Cyber-SiKu) [reply curvefs/src/tools/status/curvefs_status_base_tool.h:72]: > why some members have a default value and other not?

uri has deafult value, if you don't want get this value set it to deafult value.

File: curvefs/src/tools/curvefs_tool_define.cpp (hunk 15 above)
Comment (Cyber-SiKu) [reply curvefs/src/tools/curvefs_tool_define.cpp]: > why an extra ", " in the end?

fix

**Review by SeanHai — APPROVED** [review]

**Review by wu-hanqing — APPROVED** [review]

**Pull request merged**